<a href="https://colab.research.google.com/github/Tasneem-Ibrahim/Python/blob/main/PROJECTS_53_self_coding_challenges_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **PROJECTS**


# **53_self_coding_challenges**


## 00_intro_python


### 01_add_two_numbers

Python program that prompts the user to enter two integers, calculates their sum, and displays the result. The main() function guides the user through entering the numbers and prints their total sum with a message.





In [ ]:
# 01_add_two_numbers

def main():
    # Prompt the user for two numbers and convert inputs to integers
    num1 = int(input("Enter first number: "))
    num2 = int(input("Enter second number: "))

    # Calculate the sum
    total = num1 + num2

    # Display the result
    print("The total is " + str(total) + ".")


# This provided line is required at the end of
# Python file to call the main() function.
if __name__ == '__main__':
    main()


Enter first number: 5
Enter second number: 3
The total is 8.


In [ ]:
# 01_add_two_numbers

def main():

    print("This program adds two numbers.")
    num1 : str = input("Enter first number: ")
    num1 : int = int(num1)
    num2 : str = input("Enter second number: ")
    num2 : int = int(num2)
    total : int = num1 + num2
    print("The total is " + str(total) + ".")

if __name__ == '__main__':
    main()


This program adds two numbers.
Enter first number: 5
Enter second number: 3
The total is 8.


### 02_agreement_bot

Python program that prompts the user to share their favorite animal, then responds with a personalized message echoing their input.

In [ ]:
# 02_agreement_bot

def main():
    # Ask the user for their favorite animal
    favorite_animal = input("What's your favorite animal? ")

    # Print the response using the user's input
    print(f"My favorite animal is also {favorite_animal}!")


# This provided line is required at the end of
# Python file to call the main() function.
if __name__ == '__main__':
    main()



What's your favorite animal? cow
My favorite animal is also cow!


### 03_fahrenheit_to_celsius

A Python program that prompts the user to input a temperature in Fahrenheit, then calculates and outputs the equivalent temperature in Celsius using the conversion formula: C = (F - 32) * 5.0/9.0. This code handles floating-point Fahrenheit inputs to provide accurate Celsius results.



In [ ]:
# 03_fahrenheit_to_celsius


def main():
  # Prompt the user for a temperature in Fahrenheit and convert input to float
  degrees_fahrenheit = float(input("Enter temperature in Fahrenheit:"))

   # Convert Fahrenheit to Celsius
  degrees_celsius = (degrees_fahrenheit - 32) * 5.0 / 9.0

    # Display the result
  print(f"Temperature: {degrees_fahrenheit}F , {degrees_celsius}C")


# Python file to call the main() function.
if __name__ == '__main__':
  main()

Enter temperature in Fahrenheit:76
Temperature: 76.0F , 24.444444444444443C


### 04_how_old_are_they

Age-Related Riddle Solver
This Python program calculates the ages of five friends (Anton, Beth, Chen, Drew, and Ethan) based on a set of clues and relationships given in the riddle. The program assigns each age to a variable and outputs their names and ages in a specific format.

Age relationships:
Anton's age is fixed at 21.

Beth is 6 years older than Anton.

Chen is 20 years older than Beth.

Drew’s age is the sum of Chen's and Anton's ages.

Ethan's age is the same as Chen's age.

The program then prints each person's name and age, formatted to match exact requirements for capitalization and punctuation.

In [ ]:
# 04_how_old_are_they

def main():
    anton : int = 21  # Anton's age is given as 21 years old
    beth : int = 6 + anton  # Beth is 6 years older than Anton, so add 6 to Anton's age to get Beth's
    chen : int = 20 + beth  # Chen is 20 years older than Beth, so add 20 to Beth's age to get Chen's
    drew  : int= chen + anton  # Drew is as old as Chen's age plus Anton's age, so add them together
    ethan : int = chen  # Ethan is the same age as Chen, so set Ethan's age equal to Chen's

   # Print out all of the ages!
    print("Anton is " + str(anton))
    print("Beth is " + str(beth))
    print("Chen is " + str(chen))
    print("Drew is " + str(drew))
    print("Ethan is " + str(ethan))


# There is no need to edit code beyond this point

if __name__ == '__main__':
    main()

Anton is 21
Beth is 27
Chen is 47
Drew is 68
Ethan is 47


### 05_triangle_perimeter

Prompt the user to enter the lengths of each side of a triangle and then calculate and print the perimeter of the triangle (the sum of all of the side lengths).

In [ ]:
# 05_triangle_perimeter

def main():
  # Get the 3 side lengths of the triangle
    side1: float = float(input("What is the length of side 1? "))
    side2: float = float(input("What is the length of side 2? "))
    side3: float = float(input("What is the length of side 3? "))
    print("The perimeter of the triangle is " + str(side1 + side2 + side3))



# This provided line is required at the end of
# Python file to call the main() function.
if __name__ == '__main__':
    main()

What is the length of side 1? 3
What is the length of side 2? 4
What is the length of side 3? 5.5
The perimeter of the triangle is 12.5


### 06_square_number

Ask the user for a number and print its square (the product of the number times itself).

Here's a sample run of the program (user input is in bold italics):

Type a number to see its square: 4

4.0 squared is 16.0

In [ ]:
# 06_square_number


def main():
  num : float = float(input("Type a number to see it's square: "))

  print(str(num) + " squared is " + str(num ** 2) )

if __name__ =='__main__':
  main()

Type a number to see it's square: 4
4.0 squared is 16.0


## 01_expressions

### 01_dice simulator

Simulate rolling two dice, three times. Prints the results of each die roll. This program is used to show how variable scope works.

In [ ]:
# 01_dicesimulator

# Import the random library which lets us simulate random things like dice!
import random

# Number of sides on each die to roll
NUM_SIDES = 6

def roll_dice():
    """
    Simulates rolling two dice and prints their total
    """
    die1 : int = random.randint(1, NUM_SIDES)
    die2 : int = random.randint(1, NUM_SIDES)
    total : int =die1 + die2
    print("Total of two dice:", total)

def main():
    die1: int = 10
    print("die1 in main() starts as : " + str(die1))
    roll_dice()
    roll_dice()
    roll_dice()
    print("die1 in main() is: " + str(die1))

if __name__ == '__main__':
      main()

# This output shows how the values in main() and roll_dice() are independent, with die1 in main() unaffected by roll_dice()'s logic.

die1 in main() starts as : 10
Total of two dice: 5
Total of two dice: 9
Total of two dice: 7
die1 in main() is: 10


### 02_e=mc2.md

A program that reads mass in kilograms from the user and calculates its energy using Einstein's formula
𝐸
=
𝑚
×
𝐶
2
E=m×C
2
 , where
𝐶
=
299
,
792
,
458

𝑚
/
𝑠
C=299,792,458m/s.

In [ ]:
# 02_e=mc2.md (e)

C : int = 2997792458 # the speed of light in m/s

def main():
   mass_in_kg: float = float(input("Enter kilos of mass:"))
   # Calculate energy
   # Equivalently energy = mass*(C**2)
   # using the ** operator to raise C to the power of 2
   energy_in_joules: float = mass_in_kg * (C ** 2)
   # Display the result
   print("E = m *C^2...")
   print("m =" + str(mass_in_kg) + "kg")
   print("C = " + str(C) +"m/s")

   print(str(energy_in_joules) +" Joules of energy!")

if __name__ == '__main__':
   main()


Enter kilos of mass:8
E = m *C^2...
m =8.0kg
C = 2997792458m/s
7.1894076969933455e+19 Joules of energy!


### 03_feet_to_inches

Converts feet to inches. Feet is an American unit of measurement. There are 12 inches per foot. Foot is the singular, and feet is the plural.

In [ ]:
# 03_feet_to_inches

INCHES_PER_FOOT : int = 12

def main():
    feet : float = float(input("Enter number of feet:"))
    inches: float = feet * INCHES_PER_FOOT
    print("That is", inches, " inches.")

if __name__ == '__main__':
    main()


Enter number of feet:2
That is 24.0  inches.


### 04_pythagorean_theorem

A program that prompts the user for the lengths of the two perpendicular sides of a right triangle and calculates the hypotenuse using the Pythagorean theorem. The formula states that
𝐵
𝐶
2
=
𝐴
𝐵
2
+
𝐴
𝐶
2
BC
2
 =AB
2
 +AC
2
 , where the lengths of sides AB and AC are provided by the user.

In [ ]:
# 04_pythagorean_theorem

import math # Import math library so we can use the square root function

def main():
  # Get the 2 sides of the length from the user
  ab: float = float(input("Enter the length of AB: "))
  ac: float = float(input("Enter the length of AC: "))

  # Calculate te hypotenuse using the 2 sides and print
  bc: float= math.sqrt(ab**2 + ac**2)
  print("the length of BC (the hypotenuse) is: " + str(bc))

if __name__== '__main__':
    main()



Enter the length of AB: 3
Enter the length of AC: 4
the length of BC (the hypotenuse) is: 5.0


### 05_remainder_division

Ask the user for two numbers, one at a time, and then print the result of dividing the first number by the second and also the remainder of the division.

The result of this division is 1 with a remainder of 2



In [ ]:
# 05_remainder_division

def main():
  dividend: int = int(input("Please enter an integer to be divided:"))
  divisor: int = int(input("Please enter an integer to divided by:"))

  quotient: int = dividend // divisor
  remainder: int = dividend % divisor

  print("The result of this division is " + str(quotient) + " with a remainder of " + str(remainder))

if __name__ == '__main__':
    main()


Please enter an integer to be divided:76
Please enter an integer to divided by:4
The result of this division is 19with a remainder of 0


### 06_rolldice

Simulate rolling two dice, and prints results of each roll as well as the total.

In [ ]:
# 06_rolldice

import random
NUM_SIDES: int = 6

def main():
  die1: int = random.randint(1, NUM_SIDES)
  die2: int = random.randint(1, NUM_SIDES)
  total: int = die1 + die2

  print("Dice have ", NUM_SIDES, " sides each. ")
  print("First die:", die1)
  print("Second die:", die2)
  print("Total of two dice: ", total)

if __name__ == '__main__':
  main()

Dice have  6  sides each. 
First die: 3
Second die: 6
Total of two dice:  9


### 06_seconds_in_year

Use Python to calculate the number of seconds in a year, and tell the user what the result is in a nice print statement that looks like this (of course the value 5 should be the calculated number instead):

There are 5 seconds in a year!

You should use constants for this exercise -- there are 365 days in a year, 24 hours in a day, 60 minutes in an hour, and 60 seconds per minute.

In [ ]:
# 06_seconds_in_year

DAYS_PER_YEAR: int = 36
HOURS_PER_DAY: int = 24
MIN_PER_HOUR: int = 60
SEC_PER_MIN: int = 60

def main():
  print("There are " + str(DAYS_PER_YEAR * HOURS_PER_DAY * MIN_PER_HOUR * SEC_PER_MIN) + " seconds in a year!")

if __name__ == '__main__':
  main()




There are 3110400 seconds in a year!


### 07_tiny_mad_lib

A program that prompts the user for an adjective, a noun, and a verb, then constructs and prints a fun sentence using those words to create an entertaining story based on the Mad Libs game format.

In [ ]:
# 07_tiny_mad_lib

SENTENCE_START: str = "Panaversity is fun, I learned to program and used Python to make my " # adjective noun verb

def main():
    adjective: str =(input("Please type an adjective and press enter. "))
    noun: str = (input("Please type an noun and press enter. "))
    verb: str = (input("Please type an verb and press enter. "))

    print(SENTENCE_START + adjective + " " + noun + " " + verb + "!")

if __name__ == '__main__':
    main()


Please type an adjective and press enter. tiny
Please type an noun and press enter. plant
Please type an verb and press enter. fly
Panaversity is fun, I learned to program and used Python to make my tiny plant fly!


## 02_lists

### 01_add_many_number

Function that takes a list of numbers and returns the sum of those numbers.

In [ ]:
# 01_add_many_number

def sum_of_numbers(numbers):
  return sum(numbers)
def main():
 numbers = [1,2,3,4,5]
 result = sum_of_numbers(numbers)
 print("The sum of the numbers is:",result)

if __name__ == '__main__':
  main()

The sum of the numbers is: 15


In [ ]:
# 01_add_many_number

def add_many_numbers(numbers)-> int:
    """
    Takes in a list of numbers and returns the sum of those numbers.
    """

    total_so_far: int = 0
    for number in numbers:
        total_so_far += number

    return total_so_far

def main():
    numbers: list[int] = [1, 2, 3, 4, 5]  # Make a list of numbers
    sum_of_numbers: int = add_many_numbers(numbers)  # Find the sum of the list
    print(sum_of_numbers)  # Print out the sum above


if __name__ == '__main__':
    main()

15


### 02_double_list

program that doubles each element in a list of numbers. For example, if you start with this list:

numbers = [1, 2, 3, 4]

You should end with this list:

numbers = [2, 4, 6, 8]

In [ ]:
# 02_double_list

def double_elements(numbers):
    return [num * 2 for num in numbers]

# Example usage
numbers = [1, 2, 3, 4]
doubled_numbers = double_elements(numbers)
print("Doubled list:", doubled_numbers)


Doubled list: [2, 4, 6, 8]


In [ ]:
# 02_double_list

def main():
  numbers: list[int] = [1,2,3,4]
  for i in range(len(numbers)):
    elements_at_index = numbers[i]
    numbers[i] = elements_at_index * 2
  print(numbers)

if __name__ == '__main__':
  main()

[2, 4, 6, 8]


### 03_erase_canvas

Implement an 'eraser' on a canvas.

The canvas consists of a grid of blue 'cells' which are drawn as rectangles on the screen. We then create an eraser rectangle which, when dragged around the canvas, sets all of the rectangles it is in contact with to white.


This Code is smoothly run in VScode

In [ ]:
# 03_erase_canvas
# This Code is smoothly run in VScode

import tkinter as tk
import time

CANVAS_WIDTH = 400
CANVAS_HEIGHT = 400
CELL_SIZE = 40
ERASER_SIZE = 20

def erase_objects(canvas, eraser, eraser_position):
    """Erase objects in contact with the eraser."""
    # Define eraser boundaries
    left_x, top_y = eraser_position
    right_x = left_x + ERASER_SIZE
    bottom_y = top_y + ERASER_SIZE

    # Find overlapping objects
    overlapping_objects = canvas.find_overlapping(left_x, top_y, right_x, bottom_y)

    # Set overlapping objects' color to white (excluding eraser itself)
    for obj in overlapping_objects:
        if obj != eraser:
            canvas.itemconfig(obj, fill='white')

def main():
    # Initialize tkinter
    root = tk.Tk()
    root.title("Eraser Canvas")

    # Set up the canvas
    canvas = tk.Canvas(root, width=CANVAS_WIDTH, height=CANVAS_HEIGHT, bg='white')
    canvas.pack()

    num_rows = CANVAS_HEIGHT // CELL_SIZE
    num_cols = CANVAS_WIDTH // CELL_SIZE

    # Create a grid of blue squares
    for row in range(num_rows):
        for col in range(num_cols):
            left_x = col * CELL_SIZE
            top_y = row * CELL_SIZE
            right_x = left_x + CELL_SIZE
            bottom_y = top_y + CELL_SIZE
            # Create a cell in the grid
            canvas.create_rectangle(left_x, top_y, right_x, bottom_y, fill='blue', outline='blue')

    # Set up the eraser as a pink rectangle
    eraser = canvas.create_rectangle(0, 0, ERASER_SIZE, ERASER_SIZE, fill='pink', outline='pink')

    # Function to move eraser with the mouse
    def move_eraser(event):
        # Move the eraser to follow the mouse
        canvas.coords(eraser, event.x, event.y, event.x + ERASER_SIZE, event.y + ERASER_SIZE)
        erase_objects(canvas, eraser, (event.x, event.y))

    # Bind mouse movement to move_eraser function
    canvas.bind('<Motion>', move_eraser)

    # Run the Tkinter main loop
    root.mainloop()

if __name__ == '__main__':
    main()


### 04_flowing_with_data_structures

Using a variable storing a number as an example of scope. We saw that changes we made to the number inside a function did not stay unless we returned it. This is true for what we call immutable data types which include things like numbers and strings.

 To illustrate this, implement the add_three_copies(...) function to add three copies of a given item to a list without returning it, and compare the behavior to the x = change(x) example.

In [ ]:
# 04_flowing_with_data_structures

def add_three_copies(my_list, data):
  for i in range(3):
    my_list.append(data)

def main():
  message = input("Enter a message to copy: ")
  my_list = []
  print("List before: " , my_list)
  add_three_copies(my_list, message)
  print("List after: ", my_list)

if __name__ == '__main__':
  main()


Enter a message to copy: Best
List before:  []
List after:  ['Best', 'Best', 'Best']


### 05_get_first_element

Function get_first_element(lst) which takes in a list lst as a parameter and prints the first element in the list. The list is guaranteed to be non-empty. We've written some code for you which prompts the user to input the list one element at a time.

In [ ]:
# 05_get_first_element

def get_first_element(lst):
  """
  Print the first element in a provided list.
  """

  print(lst[0])

def get_lst():
   """
   Prompts the user to enter one element of the list at a time
   and returns the  resulting list.
   """
   lst = []
   elem: str = input("Please enter an element of the list or press enter to stop ")
   while elem !="":
    lst.append(elem)
    elem = input("Please enter an element of the list or press enter to stop ")
    return lst

def main():
  lst = get_lst()
  get_first_element(lst)

if __name__ == '__main__':
  main()

Please enter an element of the list or press enter to stop Best
Please enter an element of the list or press enter to stop Book
Best


### 06_get_last_element

Fill out the function get_last_element(lst) which takes in a list lst as a parameter and prints the last element in the list. The list is guaranteed to be non-empty, but there are no guarantees on its length.

In [ ]:
# 06_get_last_element

def get_last_element(lst):
    """
    Print the last element in a provided list.
    """
    # Takes the length of the list and subtracts 1 since they are zero-indexed (we start counting at 0)
    print(lst[len(lst) - 1])

    # The line below works too!!
    print(lst[-1])

def get_lst():
    """ Prompts the user to enter one element of the list at a time and returns the resulting list. """
    lst = []
    elem = input("Please enter an element of the list or press enter to stop. ")
    while elem != "":
        lst.append(elem)
        elem = input("Please enter an element of the list or press enter to stop. ")
    return lst

def main():
    lst = get_lst()
    get_last_element(lst)

if __name__ == '__main__':
    main()


Please enter an element of the list or press enter to stop. apple
Please enter an element of the list or press enter to stop. banana
Please enter an element of the list or press enter to stop. cherry
Please enter an element of the list or press enter to stop. 
cherry
cherry


### 07_get_list

Program which continuously asks the user to enter values which are added one by one into a list. When the user presses enter without typing anything, print the list.

Here's a sample run (user input is in blue):

Enter a value: 1 Enter a value: 2 Enter a value: 3 Enter a value: Here's the list: ['1', '2', '3']

In [ ]:
# 07_get_list

def main():
  lst = []                                # Make an empty list to store things in

  val = input("Enter a value:")           # Get an initial value
  while val:                              # While the user input isn't an empty value
      lst.append(val)                     # Add value to list
      val = input("Enter a value:")       # Get the next value to add

  print("Here's the list:", lst)

if __name__ == '__main__':
  main()

Enter a value:5
Enter a value:6
Enter a value:7
Enter a value:8
Enter a value:9
Enter a value:
Here's the list: ['5', '6', '7', '8', '9']


### 08_shorten

Fill out the function shorten(lst) which removes elements from the end of lst, which is a list, and prints each item it removes until lst is MAX_LENGTH items long. If lst is already shorter than MAX_LENGTH you should leave it unchanged. We've written a main() function for you which gets a list and passes it into your function once you run the program. For the autograder to pass you will need MAX_LENGTH to be 3, but feel free to change it around to test your program.

In [ ]:
# 08_shorten

MAX_LENGTH : int = 3

def shorten(lst):
  while len(lst) > MAX_LENGTH:
    last_elem = lst.pop()
    print(last_elem)

def get_lst():
  """
  Prompts the user to enter one element of the list at a time and returns the resulting list.
  """
  lst = []
  elem = input("Please enter an element of the list or press enter to stop. ")
  while elem != "":
     lst.append(elem)
     elem = input("Please enter an element of the list or press enter to stop. ")
  return lst

def main():
    lst = get_lst()
    shorten(lst)


if __name__ == '__main__':
    main()

Please enter an element of the list or press enter to stop. apple
Please enter an element of the list or press enter to stop. banana
Please enter an element of the list or press enter to stop. cherry
Please enter an element of the list or press enter to stop. mango
Please enter an element of the list or press enter to stop. strawberry
Please enter an element of the list or press enter to stop. 
strawberry
mango


## 03_if_statements

### 01_print_events

program that prints the first 20 even numbers. There are several correct approaches, but they all use a loop of some sort. Do no write twenty print statements

The first even number is 0:

0 2 4 6 8 10 12 14 16 18 20 22 24 26 28 30 32 34 36 38

In [ ]:
# 01_print_events

def main():
   # This for-loop start at 0 and counts up to 19 (for a total of 20 numbers)
    for i in range(20):
      print (i * 2)  # Use the 'i' value inside the for-loop

if __name__ == '__main__':
    main()

0
2
4
6
8
10
12
14
16
18
20
22
24
26
28
30
32
34
36
38


### 02_international_voting_age

Ask the user for their age and inform them if they can vote in three fictional countries with voting ages of 16, 25, and 48, respectively.

In [ ]:
PETURKSBOUIPO_AGE : int = 16
STANLAU_AGE : int = 25
MAYENGUA_AGE : int = 48

def main():
    # Get the user's age
    user_age = int(input("How old are you? "))

    # Check if the user can vote in Peturksbouipo
    if user_age >= PETURKSBOUIPO_AGE:
        print("You can vote in Peturksbouipo where the voting age is " + str(PETURKSBOUIPO_AGE) + ".")
    else:
        print("You cannot vote in Peturksbouipo where the voting age is " + str(PETURKSBOUIPO_AGE) + ".")

    # Check if the user can vote in Stanlau
    if user_age >= STANLAU_AGE:
        print("You can vote in Stanlau where the voting age is " + str(STANLAU_AGE) + ".")
    else:
        print("You cannot vote in Stanlau where the voting age is " + str(STANLAU_AGE) + ".")

    # Check if user can vote in Mayengua
    if user_age >= MAYENGUA_AGE:
        print("You can vote in Mayengua where the voting age is " + str(MAYENGUA_AGE) + ".")
    else:
        print("You cannot vote in Mayengua where the voting age is " + str(MAYENGUA_AGE) + ".")


# There is no need to edit code beyond this point

if __name__ == '__main__':
    main()

How old are you? 8
You cannot vote in Peturksbouipo where the voting age is 16.
You cannot vote in Stanlau where the voting age is 25.
You cannot vote in Mayengua where the voting age is 48.


### 03_leap_year

This program checks if a given year is a leap year based on the following rules:

The year must be divisible by 4.
If divisible by 100, it must also be divisible by 400 to be a leap year.
Using these rules, the program reads a year from the user and prints "That's a leap year!" if it qualifies, or "That's not a leap year" if it doesn’t.

In [ ]:
# 03_leap_year

def main():
  year = int(input("Enter a year: "))

  if year % 4 ==0:                            # Checking whether the provided year is evenly divisibly by 4
    if year % 100 == 0:                       # Checking whether the provided year is evenly divisibly by 100

      if year % 400 == 0:
        print("That is a leap year!")
      else:
        print("That is not a leap year!")
    else:
      print("That is a leap year!")
  else:
    print("That is not a leap year!")

if __name__ == '__main__':
  main()

### 04_tall_enough_to_ride

Program prompts user for their height and checks if it meets a 50-unit minimum; if so, it prints “You're tall enough to ride!” Otherwise, it encourages the user to try again next year.

For an additional challenge, it can keep asking for height until no input is given, then stops.

In [ ]:
# 04_tall_enough_to_ride


MINIMUM_HEIGHT : int = 50 # arbitrary units :)

def main():
  height = float(input("How tall are you? "))
  if height >= MINIMUM_HEIGHT:
    print("You are tall enough to ride.")
  else:
    print("You are not tall enough to ride, but maybe next year!")

if __name__ == '__main__':
  main()


How tall are you? 60
You are tall enough to ride.


### 05_random_numbers


Generate and print 10 random numbers between 1 and 100, with a different sequence each run. Use randint from the random library for inclusive random values.

value = random.randint(1, 6)

In [ ]:
# 05_random_numbers

import random

N_NUMBERS : int = 10
MIN_VALUE : int = 1
MAX_VALUE : int = 100

def main():

  for i in range(N_NUMBERS):
    print(random.randint(MIN_VALUE, MAX_VALUE))



if __name__ == '__main__':
  main()




27
92
53
16
21
89
22
17
98
8


## 04_dictionaries

### 00_count_nums

This program counts occurrences of each number in a list using a dictionary. Example:
Input numbers: 3, 4, 3, 6, 4, 3, 12
Output: "3 appears 3 times. 4 appears 2 times. 6 appears 1 time. 12 appears 1 time."

In [ ]:
# 00_count_nums

def get_user_numbers():
    """
    Create an empty list.
    Ask the user to input numbers and store them in a list.
    Once they enter a blank line, break out of the loop and return the list.
    """

    user_numbers = []
    while True:
        user_input = input("Enter a number :")

       # If the user enters a blank line, break out of the loop and stop asking for input
        if user_input == "":
            break

      # convert the user input to an integer and add it to the list
        num = int(user_input)
        user_numbers.append(num)

    return user_numbers

def count_nums(num_lst):
    """
    Create an empty dictionary.
    Loop over the list of numbers.
    If the number is not in the dictionary, add it as a key with a value of 1.
    If the number is in the dictionary, increment its value by 1.
    """
    num_dict = {}
    for num in num_lst:
        if num not in num_dict:
          num_dict[num] = 1
        else:
          num_dict[num] += 1
    return num_dict

def print_counts(num_dict):
    """
    Loop over the dictionary and print out each key and its value.
    """
    for num in num_dict4:
        print(str(num) + " appears " + str(num_dict[num]) + " times.")


def main():
  """
  Ask the user to input numbers and store them in a list. Once they print
  out the numbers of times each numbers appeeas in the list.
  """
  user_numbers = get_user_numbers()
  num_dict = count_nums(user_numbers)
  print_counts(num_dict)

  # Python boilerplate
if __name__ == '__main__':
    main()

Enter a number :3
Enter a number :4
Enter a number :5
Enter a number :4
Enter a number :3
Enter a number :12
Enter a number :
3 appears 2 times.
4 appears 2 times.
5 appears 1 times.
12 appears 1 times.


### 01_phonebook

In this program we show an example of using dictionaries to keep track of information in a phonebook.

In [ ]:
# 01_phonebook

def read_phone_numbers():
  """
  Ask the user for names/numbers to story in a phonebook (dictionary).
  Returns the phonebook.
  """
  phonebook = {}

  while True:
    name = input("Name: ")
    if name == "":
      break
    number = input("Number: ")
    phonebook[name] = number

  return phonebook

def print_phonebook(phonebook):
  """
  Prints out all the names and numbers in the phonebook.
  """
  for name in phonebook:
    print(str(name) + " -> " + str(phonebook[name]))

def lookup_numbers(phonebook):
  """
  Allow the user to look up phonenumbers in the phonebook
  by looking up the number associated with a name.
  """
  while True:
    name = input("Enter name to lookup: ")
    if name == "":
      break
    if name not in phonebook:
      print(name + " is not in the phonebook. ")
    else:
      print(phonebook[name])

def main():
  phonebook = read_phone_numbers()
  print_phonebook(phonebook)
  lookup_numbers(phonebook)

  # Python boilerplate
if __name__ == '__main__':
    main()

Name: Tasneem
Number: 0333
Name: Areej
Number: 0987
Name: Fatima
Number: 6543
Name: 
Tasneem -> 0333
Areej -> 0987
Fatima -> 6543
Enter name to lookup: Tasneem
0333
Enter name to lookup: Areej
0987
Enter name to lookup: Fatima
6543
Enter name to lookup: 


### 02_pop_up_shop

Write a program that calculates the total cost of fruit based on quantities entered by the user from a given price list. Prompt for each fruit’s quantity, then print the combined cost.

In [ ]:
# 02_pop_up_shop

def main():
    fruits = {'apple': 1.5, 'durian': 50, 'jackfruit': 80, 'kiwi': 1, 'rambutan': 1.5, 'mango': 5}

    total_cost = 0
    for fruit_name in fruits:
        price = fruits[fruit_name]
        amount_bought = int(input("How many (" + fruit_name + ") do you want to buy?: "))
        total_cost += (price * amount_bought)

    print("Your total is $" + str(total_cost))


# There is no need to edit code beyond this point

if __name__ == '__main__':
    main()

### 03_powerful_passwords

Program that verifies passwords for websites by comparing stored password hashes, without storing actual passwords. Use the provided hash_password(...) function to hash password_to_check and return True if it matches the stored hash for the given email in stored_logins. (Hint: hash_password(...) returns a password hash.)

In [ ]:
# 03_powerful_passwords

from hashlib import sha256

def login(email, stored_logins, password_to_check):
    """
    Returns True if the hash of the password we are checking matches the one in stored_logins
    for a specific email. Otherwise, returns False.

    email: the email we are checking the password for
    stored_logins: a dictionary pointing from an email to its hashed password
    password_to_check: a password we want to test alongside the email to login with
    """

    if stored_logins[email] == hash_password(password_to_check):
        return True

    return False

# There is no need to edit code beyond this point

def hash_password(password):
    """
    Takes in a password and returns the SHA256 hashed value for that specific password.

    Inputs:
        password: the password we want

    Outputs:
        the hashed form of the input password
    """

    return sha256(password.encode()).hexdigest()

def main():
    # stored_logins is a dictionary with emails as keys and hashed passwords as values
    stored_logins = {
        "example@gmail.com": "5e884898da28047151d0e56f8dc6292773603d0d6aabbdd62a11ef721d1542d8",
        "code_in_placer@cip.org": "973607a4ae7b4cf7d96a100b0fb07e8519cc4f70441d41214a9f811577bb06cc",
        "student@stanford.edu": "882c6df720fd99f5eebb1581a1cf975625cea8a160283011c0b9512bb56c95fb"
    }

    print(login("example@gmail.com", stored_logins, "word"))
    print(login("example@gmail.com", stored_logins, "password"))

    print(login("code_in_placer@cip.org", stored_logins, "Karel"))
    print(login("code_in_placer@cip.org", stored_logins, "karel"))

    print(login("student@stanford.edu", stored_logins, "password"))
    print(login("student@stanford.edu", stored_logins, "123!456?789"))


if __name__ == '__main__':
    main()

## 05_loops_control_flow

### 00_guess_my_number

Guess My Number" game where you guess a number between 0 and 99. The program gives feedback whether the guess is too high, too low, or correct, until you guess the right number.

In [ ]:
import random

def main():
    # Generate the secret number at random!
    secret_number = random.randint(1, 99)

    print("I am thinking of a number between 1 and 99...")

    # Get user's guess
    guess = int(input("Enter a guess: "))
    # True if guess is not equal to secret number
    while guess != secret_number:
        if guess < secret_number:  # If-statement is True if guess is less than secret number
            print("Your guess is too low")
        else:
            print("Your guess is too high")

        print() # Print an empty line to tidy up the console for new guesses
        guess = int(input("Enter a new guess: "))  # Get a new guess from the user

    print("Congrats! The number was: " + str(secret_number))

if __name__ == '__main__':
    main()

I am thinking of a number between 1 and 99...
Enter a guess: 5
Your guess is too low

Enter a new guess: 9
Your guess is too low

Enter a new guess: 78
Your guess is too high

Enter a new guess: 40
Your guess is too low

Enter a new guess: 48
Your guess is too low

Enter a new guess: 50
Your guess is too low

Enter a new guess: 60
Your guess is too high

Enter a new guess: 55
Your guess is too high

Enter a new guess: 51
Congrats! The number was: 51


### 01_fibonacci

Write a program that prints Fibonacci sequence terms starting from 0 and 1, continuing until the terms exceed 10,000. The program should display the terms as follows: 0 1 1 2 3 5 8 13 21 34 55 89 144 233 377 610 987 1597 2584 4181 6765.

In [ ]:
# 01_fibonacci

MAX_TERM_VALUE : int = 1000

def main():
    curr_term = 0 # The 0th Fibonacci Number
    next_term = 1 # The 1st Fibonacci Number
    while curr_term <= MAX_TERM_VALUE:
        print(curr_term)
        term_after_next = curr_term + next_term
        curr_term = next_term
        next_term = term_after_next
if __name__ == '__main__':
    main()
#

0
1
1
2
3
5
8
13
21
34
55
89
144
233
377
610
987


### 02_print_events

A program that prints the first 20 even numbers. There are several correct approaches, but they all use a loop of some sort.

The first even number is 0:

0 2 4 6 8 10 12 14 16 18 20 22 24 26 28 30 32 34 36 38



In [ ]:
# 02_print_events

def main():
    # This for-loop start at 0 and counts up to 19 (for a total of 20 numbers)
    for i in range(20):
        print(i * 2)  # Use the 'i' value inside the for-loop

# Call the main function when "run", no need to edit anything below!
if __name__ == "__main__":
    main()

0
2
4
6
8
10
12
14
16
18
20
22
24
26
28
30
32
34
36
38


### 03_wholesome_machine

 A program that repeatedly prompts the user to type the affirmation "I am capable of doing anything I put my mind to." until they type it correctly. The program will acknowledge the correct input with "That's right! :)" after the user enters it.








In [ ]:
# 03_wholesome_machine

AFFIRMATION : str = "I am capable of doing anything I put my mind to."

def main():
  print("Please type the following affirmation: " + AFFIRMATION)

  user_feedback = input() # Get user's input
  while user_feedback != AFFIRMATION:   # While the user's input isn't the affirmation

    # Tell the user that they did not type the affirmation correctly
    print("That was not the affirmation.")

    # Ask the user to type the affirmation again!
    print("Please type the following affirmation: " + AFFIRMATION)
    user_feedback = input()

  print("That's right! :)")

if __name__ == '__main__':
  main()



Please type the following affirmation: I am capable of doing anything I put my mind to.
I am capable of doing anything I put my mind to.
That's right! :)


### 04_liftoff
A program that counts down from 10 to 1 and then prints "Liftoff!" using a for loop. The loop will print the numbers from 10 to 1 in descending order, followed by the message "Liftoff!".

In [ ]:
# 04_liftoff

def main():
    # Countdown from 10 to 1
    for i in range(10, 0, -1):  # Start from 10, go down to 1, decrement by 1 each time
        print(i, end=" ")       # Print the countdown numbers on the same line with a space
    print("Liftoff!")           # Print "Liftoff!" after the countdown is complete

# This line is required at the end of the Python file to call the main() function.
if __name__ == '__main__':
    main()


10 9 8 7 6 5 4 3 2 1 Liftoff!


### 05_double_it

A program that asks the user to enter a number, then repeatedly doubles it until the value reaches 100 or more. The program prints the results at each step. Use a while loop to continue doubling the value as long as it is less than 100.








In [ ]:
# 05_double_it

def main():
    # Ask the user to enter a number
    curr_value = int(input("Enter a number: "))

    # Use a while loop to double the number until it's 100 or more
    while curr_value < 100:
        curr_value = curr_value * 2  # Double the current value
        print(curr_value, end=" ")   # Print the doubled value on the same line

# This line is required at the end of the Python file to call the main() function.
if __name__ == '__main__':
    main()


Enter a number: 5
10 20 40 80 160 

## 06_functions

### 00_averages

A function that takes two numbers and finds the average between the two.



In [ ]:
# 00_averages

def average(a: float, b: float):
  """
  Return the number which is half way between a and b
  """
  sum = a + b
  return sum / 2

def main():
  avg_1 = average(0, 10)
  avg_2 = average(8, 10)

  final = average(avg_1, avg_2)
  print("avg_1", avg_1)
  print("avg_2", avg_2)
  print("final", final)

if __name__ == '__main__':
    main()

avg_1 5.0
avg_2 9.0
final 7.0


### 01_chaotic_counting

Create a function chaotic_counting() that prints numbers 1 to 10 but stops randomly based on done() returning True. If it stops early, control returns to main(), which prints "I'm done." Example: "1 2 3 I'm done."

In [ ]:
# 01_chaotic_counting

import random

# Define the likelihood of stopping early (adjust this between 0 and 1 if needed)
DONE_LIKELIHOOD = 0.3

def chaotic_counting():
  for i in range(10):
    curr_num = i + 1
    if done():
        return # this ends the function execution, so we'll get back to the main()
    print(curr_num)

def done():
  """ Returns True with a probability of DONE_LIKELIHOOD """
  if random.random() < DONE_LIKELIHOOD:
    return True
  return False

def main():
  print("I'm going to count until 10 or until I feel like stopping, whichever comes first.")
  chaotic_counting()
  print("I'm done")

if __name__ == '__main__':
  main()

I'm going to count until 10 or until I feel like stopping, whichever comes first.
I'm done


### 02_count_even

Define count_even(lst) to:

Prompt the user to enter integers until pressing enter.
Count and print the number of even numbers in the list.
Use the provided input prompt: "Enter an integer or press enter to stop: ".

In [ ]:
# 02_count_even

def count_even(lst):
  """
  Returns number of even numbers in the list.
  >>> count_even([1, 2, 3, 4])
  2
  >>> count_even([1, 3, 5, 7])
  0
  """

  count = 0
  for num in lst:
    if num % 2 == 0:
      count += 1

 # Here's another way to do this same thing, with a different kind of for-loop:
    # for i in range(len(lst)):
    #     num = lst[i]
    #     if num % 2 == 0:
    #         count += 1

    print(count)

def get_list_of_ints():
    """
    Read in integers until the user pressess and return the resulting list
    """
    lst = []
    user_input = input("Enter an integer or press enter to stop: ")  # Get user input for an integer
    while user_input != "":  # While the user doesn't enter nothing...
        lst.append(int(user_input))  # Cast the user input into an integer and add it to our list
        user_input = input("Enter an integer or press enter to stop: ")  # Get the next user input

    return lst

    def main():
      lst = get_list_of_ints()
      count_even(lst)

if __name__ == '__main__':
    main()


I'm going to count until 10 or until I feel like stopping, whichever comes first.
1
2
3
I'm done


### 04_double

Fill out the double(num) function to return the result of multiplying num by 2. We've written a main() function for you which asks the user for a number, calls your code for double(num) , and prints the result.

Here's a sample run of the program (user input in bold italics):

Enter a number: 2 Double that is 4

In [ ]:
# 04_double

def double(num: int):
  return num * 2

def main():
  num = int(input("Enter a number: "))
  num_times_2 = double(num)
  print("Double that is", num_times_2)

if __name__ == '__main__':
  main()

Enter a number: 3
Double that is 6


### 05_get_name

Complete get_name() to return your name as a string (use "Sophia" for autograder). main() prints a greeting using this function.

In [ ]:
# 05_get_name

def get_name():
  return "Sophia"

def main():
  name = get_name()
  print("Howdy", name, "!🙂")

if __name__ == '__main__':
  main()

Howdy Sophia !🙂


### 06_is_odd

10 even 11 odd 12 even 13 odd 14 even 15 odd 16 even 17 odd 18 even 19 odd

In [ ]:
# 06_is_odd

def main():
  for i in range(10):
    if is_odd(i):
      print('odd')
    else:
      print('even')

def is_odd(value:int):
  """
  Check to see if a value is odd. if it's , resume true.
  """
  remainder = value % 2 # 0 if value is divisible by 2, 1 if it isn't
  return remainder == 1

if __name__ == '__main__':
  main()


even
odd
even
odd
even
odd
even
odd
even
odd


### 07_print_divisor

Define print_divisors(num) to print all divisors of num from 1 to num. Call it in main() to display divisors based on user input.

In [ ]:
def print_divisors(num: int):
  print("Here are the divisors of ", num )
  for i in range(num):
    curr_divisor = i + 1
    if num % curr_divisor == 0:
      print(curr_divisor)

def main():
  num = int(input("Enter a number: "))
  print_divisors(num)

if __name__ == '__main__':
  main()

Enter a number: 50
Here are the divisors of  50
1
2
5
10
25
50


### 08_print_multiple


Define print_multiple(message, repeats) to print message exactly repeats times. Use it in main() to handle user input and display the repeated message.

In [ ]:
def print_multiple(message, repeats):
    for _ in range(repeats):
        print(message)

def main():
    message = input("Please type a message: ")
    repeats = int(input("Enter a number of times to repeat your message: "))
    print_multiple(message, repeats)

if __name__ == '__main__':
    main()


Please type a message: Hello World
Enter a number of times to repeat your message: 4
Hello World
Hello World
Hello World
Hello World


### 09_sentence_generator

Define make_sentence(word, part_of_speech) to print a sentence based on part_of_speech:

If part_of_speech is 0, we will assume the word is a noun and use the template: "I am excited to add this ____ to my vast collection of them!"

If part_of_speech is 1, we will assume the word is a verb use the template: "It's so nice outside today it makes me want to ____!"

If part_of_speech is 2, we will assume the word is an adjective and use the template: "Looking out my window, the sky is big and ____!" make_sentence(word, part_of_speech) should not return anything, just print the correct sentence with the word filled in the blank.

Here's a sample run of the program (user input is in blue):

Please type a noun, verb, or adjective: groovy Is this a noun, verb, or adjective? Type 0 for noun, 1 for verb, 2 for adjective: 2 Looking out my window, the sky is big and groovy!

In [ ]:
# 09_sentence_generator

def make_sentence(word, part_of_speech):
  if part_of_speech == 0:
    # noun
    print("I am excited to add this "+ word + " to my vast collection of them!")
  elif part_of_speech == 1:
    # verb
    print ("It's so nice outside today it makes me want to " + word + "!")
  elif part_of_speech == 2:
    # adjective
    print("Looking out my window, the sky is big and " + word + "!")
  else:
    # part_of_speech is invalid (not 0, 1, or 2)
    print("Part of speech must be 0, 1, or 2! Can't make a sentence. ")
def main():
  word : str = input("Please type a noun, verb, or adjective ")
  print("Is this a noun, verb or adjective? ")
  part_of_speech =int(input("Type 0 for noun, 1 for verb, 2 for adjective "))
  make_sentence(word, part_of_speech)

if __name__ == '__main__':
  main()

Please type a noun, verb, or adjective bright
Is this a noun, verb or adjective? 
Type 0 for noun, 1 for verb, 2 for adjective 2
Looking out my window, the sky is big and bright!


### 10_print_ones_digit

A function called print_ones_digit , which takes as a parameter an integer num and prints its ones digit. The modulo (remainder) operator, %, should be helpful to you here. Call your function from main()!

Here's a sample run (user input is in blue):

Enter a number: 42 The ones digit is 2


In [ ]:
#10_print_ones_digit

def print_ones_digit(num):
  print("The ones digit is ", num % 10)

def main():
  num = int(input("Enter a number: "))
  print_ones_digit(num)


if __name__ == '__main__':
  main()

Enter a number: 368
The ones digit number is  8


## 07_information_flow

### 00_choosing_returns

Fill out is_adult(age) to return True if age >= ADULT_AGE, otherwise return False. Here's a sample:

Input: 35 → Output: True
Input: 7 → Output: False

In [ ]:
# 00_choosing_returns

ADULT_AGE : int = 18

def is_adult(age: int):
  if age >= ADULT_AGE:
    return True
  return False

def main():
  age : str = int(input("How old is the person?:"))
  print(is_adult(age))

if __name__ == '__main__':
    main()

How old is the person?:32
True


### 01_greetings


Use greet(name) to print a greeting. In main(), prompt the user for their name and call greet(name).

Sample:
Input: Sophia → Output: Greetings Sophia!

In [ ]:
# 01_greetings

def main():
  name : str = input("What's your name? ")
  print(greet(name))

def greet(name: str):
  return "Greetings " + name + "!"

if __name__ == '__main__':
  main()

What's your name? Tas
Greetings Tas!


### 02_in_range

Implement the following function which takes in 3 integers as parameters:

def in_range(n, low, high) """ Returns True if n is between low and high, inclusive. high is guaranteed to be greater than low. """

In [ ]:
def in_range(n, low, high):
    """
    Returns True if n is between low and high, inclusive.
    high is guaranteed to be greater than low.
    """
    return low <= n <= high  # Checks if n is within the inclusive range of low and high

def main():
    # Testing the in_range function with various inputs
    print("Testing in_range function:")
    print(in_range(5, 1, 10))    # Expected output: True, because 5 is between 1 and 10
    print(in_range(0, 1, 10))    # Expected output: False, because 0 is outside the range 1 to 10
    print(in_range(10, 1, 10))   # Expected output: True, because 10 is within the range 1 to 10
    print(in_range(1, 1, 10))    # Expected output: True, because 1 is within the range 1 to 10

# This line is required to call the main() function when the script is run directly
if __name__ == '__main__':
    main()

Testing in_range function:
True
False
True
True


### 03_in_stock

Sophia's fruit store has a function num_in_stock(fruit) that returns the stock of a given fruit. In main(), prompt the user to input a fruit, check its stock, Call num_in_stock(fruit) to get the number of that fruit that Sophia has in stock and print the result:

If stock > 0: print the quantity.
Otherwise: print "This fruit is not in stock."
Example Runs:

Input: pear
Output:
This fruit is in stock! Here is how many:
1000

Input: lychee
Output:
This fruit is not in stock.

In [ ]:
# 03_in_stock

def main():
  fruit : str = input("Enter a fruit: ")
  stock = num_in_stock(fruit)
  if stock == 0:
        print("This fruit is not in stock.")
  else:
        print("This fruit is in stock! Here is how many: ")
        print(stock)

def num_in_stock(fruit):
  """
	This function returns the number of fruit Sophia has in stock.
	"""
  if fruit == 'apple':
    return 2
  if fruit == 'durian':
    return 4
  if fruit == 'pear':
    return 1000
  else:
		# this fruit is not in stock.
		    return 0

if __name__ == '__main__':
  main()

Enter a fruit: apple
This fruit is in stock! Here is how many: 
2


### 04_multiple_returns

Task: Fill out the get_user_data() function to collect user information.
Prompt the user for:

First name: Store in first_name.
Last name: Store in last_name.
Email address: Store in email.
Return all three values as a tuple: (first_name, last_name, email).

In [ ]:
# 04_multiple_returns

def get_user_data():
  first_name: str = input("What is your first name?: ")
  last_name: str = input("What is your last name?: ")
  email_adress: str = input("What is your email address?: ")
  return first_name, last_name, email_address

def main():
    user_data = get_user_info()
    print("Received the following user data:", user_data)

if __name__ == '__main__':
  main()

What is your first name?: Tas
What is your last name?: Ibm
What is your email address?: tas@gamil.com
Received the following user data: ('Tas', 'Ibm', 'tas@gamil.com')


### 05_subtract_7

Fill out the subtract_seven helper function to subtract 7 from num, and fill out the main() method to call the subtract_seven helper function! If you're stuck, revisit the add_five example from lecture.

In [ ]:
# 05_subtract_7

def subtract_seven(num):
    """Subtracts 7 from the given number."""
    return num - 7

def main():
    # Example usage of subtract_seven
    num = int(input("Enter a number: "))
    result = subtract_seven(num)
    print(f"Result after subtracting 7: {result}")

# This provided line is required at the end of
# Python file to call the main() function.
if __name__ == '__main__':
    main()


Enter a number: 3
Result after subtracting 7: -4


In [ ]:
# 05_subtract_7

def main():
	num: int = 7
	num = subtract_seven(num)
	print("this should be zero: ", num)

def subtract_seven(num):
	num = num - 7
	return num


# There is no need to edit code beyond this point

if __name__ == '__main__':
    main()

this should be zero:  0


Enter a number: 4
Result after subtracting 7: -3
